### PR TITLE
feat(ios): add support for ios 16

### LINKED ISSUE #100: iOS 16 may not respect a request to change orientation
# Bug Report

## Problem

Getting this issue on our radar early:

https://developer.apple.com/forums/thread/707735?answerId=721941022#721941022

### What is expected to happen?

This approach used to work, but may not on iOS 16:
 [[UIDevice currentDevice] setValue:@(orientation) forKey:@"orientation"]

We may need to switch to:
https://developer.apple.com/documentation/uikit/uiwindowscene/3975944-requestgeometryupdate/
(which would require building from XCode 14)

### Version information

iOS 16 beta 4

## Checklist

- [ ] I searched for existing GitHub issues
- [ ] I updated all Cordova tooling to most recent version
- [ ] I included all the necessary information above

### PR DESCRIPTION
**Pull request by oofaish**
### Platforms affected
Apple devices running iOS 16

### Motivation and Context

resolves #100  

Currently orientation is broken on iOS 16 with
`Setting UIDevice.orientation is not supported. Please use UIWindowScene.requestGeometryUpdate(_:)`

### Description

deprecate on whether iOS 16 is available - either use the old method or the new one.

### Testing

confirmed on iPhone 11 running iOS 15 and iPhone 13 Pro running iOS 16.
I have not confirmed all features work, but simply setting orientations works - the old pre iOS 16 should be unaffected.

### Checklist

- [x] I've run the tests to see all new and existing tests pass
- [ ] I added automated test coverage as appropriate for this change
- [x] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
- [x] I've updated the documentation if necessary

### CODE REVIEW

File: src/ios/CDVOrientation.m
Comment (guillaumeperaldi) [inline src/ios/CDVOrientation.m:83]: Hi, it seems that your fix is not working.
I tried on my app, and when I use your code I have this error:
_[Orientation] BUG IN CLIENT OF UIKIT: Setting UIDevice.orientation is not supported. Please use UIWindowScene.requestGeometryUpdate(_:)_

I'm not fluent to this langage but I guess that the problem is around there:
[[UIDevice currentDevice] setValue:value forKey:@"orientation"]; will not work with iOs16.

Thanks :)

Comment (guillaumeperaldi) [reply src/ios/CDVOrientation.m:83]: Hi, yes i'm using iphone 16, version 16.0 (20A362) on iPhone8.
The code patch in my app seems ok too: cordova / ionic / angular.
The context: my app is locked in portrait mode, until i need to rotate the screen by code to landscape mode to display a chart.
Do you need more informations? Hope it can help.

Comment (oofaish) [reply src/ios/CDVOrientation.m:83]: how are you adding the plugin? maybe remove the plugin:
`ionic cordova plugin remove cordova-plugin-screen-orientation`
`ionic cordova plugin add https://github.com/oofaish/cordova-plugin-screen-orientation.git#ios-16`

if you already have the right version, then I suggest putting a debuger in xcode in `src/ios/CDVOrientation.m`  and just stepping through to see where it goes wrong.

Comment (guillaumeperaldi) [reply src/ios/CDVOrientation.m:83]: Hi, finally, we checked our CI and the way we build our app, one thing was missing: rm -rf plugins so the ipa was based on the old plugin, not on the patch.
Everything is ok, thanks for your job, it works perfectly, you were right. :)

Comment (webfletch) [reply src/ios/CDVOrientation.m:83]: I have tried adding the plugin as you described (i.e. remove and then add the iOS 16 version) but when trying to build I am getting various errors such as the following, any idea what could be causing this?

`use of undeclared identifier 'UIWindowSceneGeometryPreferencesIOS'`

Comment (oofaish) [reply src/ios/CDVOrientation.m:83]: Which version of Xcode are you using? I don’t think the latest?

Comment (webfletch) [reply src/ios/CDVOrientation.m:83]: I am building using Ionic AppFlow, the version it is using to do this looks to be 13.4.1

Comment (oofaish) [reply src/ios/CDVOrientation.m:83]: If you can’t bump that up to 14 then you ll have to deprecate based on Xcode version which my code doesn’t do. I am not sure what the ups and downs of it are and not going to have time to explore I am afraid.

Comment (webfletch) [reply src/ios/CDVOrientation.m:83]: No problem, I have got in touch with Ionic to ask when they will be adding Xcode 14, in the meantime I am adding an option for the user to be able to disable the screen rotation functionality as a setting in the App to stop people on iOS 16 from getting stuck in landscape mode when the app is generally supposed to run in portrait mode.

**Review by sdre-tech — APPROVED** [review]

File: src/ios/CDVOrientation.m
Comment (laughsky) [reply src/ios/CDVOrientation.m:83]: if use xcode 13，the following is better:

        SEL supportedInterfaceSelector = NSSelectorFromString(@"setNeedsUpdateOfSupportedInterfaceOrientations");
        [self.viewController performSelector:supportedInterfaceSelector];

**Comment by Theblood** [comment]
Please Confirm this PR!!

**Review by webfletch — APPROVED** [review]

**Review by Theblood — APPROVED** [review]

**Comment by jindrichkuba** [comment]
can u please release this change?

**Comment by webfletch** [comment]
What is holding up the release of this change?

**Comment by thmclellan** [comment]
Thanks @oofaish for sharing your fix! I've been testing on iOS 16.1 on a physical iPhone 13 Pro Max (Capacitor 4.4, Ionic 5, Angular 12, XCode 14.1, Ventura 13.0) and noticed two issues: 

1 - Starting from primary portrait without my screen locked, calling `window.screen.orientation.lock("landscape")` causes a strange double rotation (see attached video). Calling window.screen.orientation.lock("landscape-primary") works as expected.  

From debugging, it seems [L68](https://github.com/oofaish/cordova-plugin-screen-orientation/blob/b70782c4fc118625555ab739c746e27c3ae50af1/src/ios/CDVOrientation.m#L68)  (UIInterfaceOrientationMaskLandscapeLeft) is getting called in the general landscape scenario, while L74 UIInterfaceOrientationMaskLandscapeRight is getting called on landscape-primary. 

2 - Calling `window.screen.orientation.unlock()` is causing a requestGeometryUpdate error in Xcode: 

`2022-11-08 11:44:59.124078-0800 App[3407:1282732] [Orientation] BUG IN CLIENT OF UIKIT: Setting UIDevice.orientation is not supported. Please use UIWindowScene.requestGeometryUpdate(_:)`

It seems L93 is being called on iOS 16: 
`[[UIDevice currentDevice] setValue:[NSNumber numberWithInt:_lastOrientation] forKey:@"orientation"];`

XCode complains but still executes L96 (`setNeedsUpdateOfSupportedInterfaceOrientations`) and presumably that does the unlock. 

This is minor stuff but wanted to share what I found while testing. Thanks again


https://user-images.githubusercontent.com/9682900/200661364-19f4e0f9-6cd0-4bd1-b8d7-b18dd46d0ef0.mov

**Comment by oofaish** [comment]
thanks @thmclellan   I can have a look in the next couple of days

**Comment by oofaish** [comment]
@thmclellan the latest commit in the branch fixes the double rotation issue - the code is definitely a bit messy now - but seems to be working fine on the cases I have tried.

File: package-lock.json
Comment (erisu) [inline package-lock.json:8]: Same here, this should not be changed

**Review by erisu — CHANGES_REQUESTED** [review]
Overall it might be OK, but as you suggested in a previous commit, 

> the code is definitely a bit messy now - but seems to be working fine

What if you split the code out into seperate methods?

```objectivec
-(void)handleAboveEqualIos16
{
    NSObject *value;

    if (orientationMask != 15) {
        ...
    } else {
        ...
    }

    if (value != nil) {
        _isLocked = true;
        UIWindowScene *scene = (UIWindowScene*)[[UIApplication.sharedApplication connectedScenes] anyObject];
        ...
    } else {
        _isLocked = false;
    }
}

-(void)handleBelowEqualIos15
{
    NSValue *value;

    if (orientationMask != 15) {
        ...
    } else {
        ...
    }

    if (value != nil) {
        _isLocked = true;
        [[UIDevice currentDevice] setValue:value forKey:@"orientation"];
    } else {
        _isLocked = false;
    }
}
```

And then in the `screenOrientation` write

```objectivec
if ([UIDevice currentDevice] != nil){
    if (@available(iOS 16.0, *)) {
        [self handleAboveEqualIos16];
    } else {
        [self handleBelowEqualIos15];
    }
}
```

It makes some of the code look repeative but it keeps most of the iOS 15 and below and 16 and above code seperated. Not all is split out but the main chunk of it.

Also you no longer need to have this `value` and `value16` name since its all split as well.. keep it simple as `value`

**Comment by thmclellan** [comment]
Thanks @oofaish for these changes. Tested on iOS 16 physical iPhone 13 Pro Max and the "landscape" lock is working as expected now. The "unlock" call still results in the below xcode warning but it's working fine from a user standpoint: 

`[Orientation] BUG IN CLIENT OF UIKIT: Setting UIDevice.orientation is not supported. Please use UIWindowScene.requestGeometryUpdate(_:)) `

**Comment by undimon** [comment]
`                    [scene requestGeometryUpdateWithPreferences:(UIWindowSceneGeometryPreferencesIOS*)value16 errorHandler:^(NSError * _Nonnull error) {
                        NSLog(@"Failed to change orientation  %@ %@", error, [error userInfo]);
                        // do nothing
                    }];`

I get this error:
`None of the requested orientations are supported by the view controller`

**Comment by marco-hd** [comment]
**iPhone XR / iOS 16.1.1 / XCode 14.1**

Hi guys,

I can confirm:
 - rotation is working
 - double rotation is fixed

However, the unlocking doesn't happen anymore.

I have no experience with Obj-C (please bear with me), but I tried to make it work this way:

```
...
if (_lastOrientation != UIInterfaceOrientationUnknown) {                                        
    ((void (*)(CDVViewController*, SEL, NSMutableArray*))objc_msgSend)(vc,selector,result);
    if (@available(iOS 16.0, *)) {
        [self.viewController setNeedsUpdateOfSupportedInterfaceOrientations];
        value16 = [[UIWindowSceneGeometryPreferencesIOS alloc] initWithInterfaceOrientations:_lastOrientation];
    }
    else {
        [[UIDevice currentDevice] setValue:[NSNumber numberWithInt:_lastOrientation] forKey:@"orientation"];                      
        [UINavigationController attemptRotationToDeviceOrientation];
    }
}
...
```

and the _lastOrientation initialisation:

```
...
if (!_isLocked) {
      if (@available(iOS 16.0, *)) {
          UIWindowScene *scene = (UIWindowScene*)[[UIApplication.sharedApplication connectedScenes] anyObject];

          switch(scene.interfaceOrientation) {
              case UIInterfaceOrientationUnknown:
                  _lastOrientation = UIInterfaceOrientationMaskAll;
                  break;
              case UIInterfaceOrientationPortrait:
                  _lastOrientation = UIInterfaceOrientationMaskPortrait;
                  break;
              case UIInterfaceOrientationPortraitUpsideDown:
                  _lastOrientation = UIInterfaceOrientationMaskPortraitUpsideDown;
                  break;
              case UIInterfaceOrientationLandscapeLeft:
                  _lastOrientation = UIInterfaceOrientationMaskLandscapeLeft;
                  break;
              case UIInterfaceOrientationLandscapeRight:
                  _lastOrientation = UIInterfaceOrientationMaskLandscapeRight;
                  break;
          }

      } else {
          _lastOrientation = [UIApplication sharedApplication].statusBarOrientation;        
      }      
}
...
```

So, it kind of works, but the locked screen can be unlocked by pure rotation of the device.
Any idea how to keep it locked until it is explicitly unlocked from the JS WebView world?

**Comment by VasanthCBH** [comment]
Any update about the merge of this PR or a fix for this screen orientation issue on IOS?

**Comment by oofaish** [comment]
Ooh thanks for the review @erisu. I can take a look. the changes that you say should not be there were due to another commit that was not in master it seems. Weird. I ll sort it out.

**Review by Theblood — CHANGES_REQUESTED** [review]
Please Change

**Comment by oofaish** [comment]
@erisu here is an updated PR:
- go back and fix the issue with locking not working (had to call `((void (*)(CDVViewController*, SEL, NSMutableArray*))objc_msgSend)(vc,selector,result);` even in iOS16
- only call `setNeedsUpdateOfSupportedInterfaceOrientations` at the end so we dont end up double rotating - seems to work fine 
- split the code into post ios16 and pre ios16 as requested.

File: src/ios/CDVOrientation.m
Comment (erisu) [inline src/ios/CDVOrientation.m:33]: Can you please confirm, if this iOS 16 conditional wrapper can be removed?

I believe the method is called only on line 127 and it is already wrapped by an iOS 16 conditional check (Line 126). If so we can remove the extra if conditional check and de-indent the lines.

Other thing this.. I think everything looks good now and cleaner.

Comment (erisu) [reply src/ios/CDVOrientation.m:33]: Actually can you try and replace the wrapper with this instead?

```objective-c
#if __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_15_5
// Code Here
#endif
```

Just to see if we can fix the CI.

I ran the CI and it failed, but this is because the earlier iOS versions are built with a version of XCode that doesn't understand the newer identifiers.

E.g. error message:
> error: use of undeclared identifier 'UIWindowSceneGeometryPreferencesIOS'

Comment (oofaish) [reply src/ios/CDVOrientation.m:33]: it doesn't look like I can replace it, because if I do then it fails to compile on Xcode14 (because on Xcode14 max allow is iOS16) - but I think I can use a combination of the two directives.

the wrapper on line 33 is there to stop Xcode complaining with warnings, I am trying to see if I can add a pragma just for that function but I dont seem to be able to find something sensible.

anyway new PR coming up.

**Comment by oofaish** [comment]
Have a look at latest PR see what you think @erisu  - appreciate all the comments.

**Comment by oofaish** [comment]
Hmm, looks like the tests failed again? 😡

**Comment by oofaish** [comment]
should I be using 
```
#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 160000
```
instead of
```
#if __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_15_5
```

**Comment by erisu** [comment]
> should I be using
> 
> ```
> #if __IPHONE_OS_VERSION_MAX_ALLOWED >= 160000
> ```
> 
> instead of
> 
> ```
> #if __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_15_5
> ```
> 
> confirmed if I use `>= 170000` on xCode vs `>= 160000` it will switch branches in xCode14, but not with `__IPHONE_15_5` vs `__IPHONE_16_5`, not the perfect test, but there it is

Awesome, the update fixes the CI and everything is now green.

The changes look great, clean, and readable to me.

The only other small nitpick that I had was the variable name `value16` could be renamed to `value`. Since the code was split out per iOS version, there are no more overlapping variable name usages.

But, other than that, everything is great.

**Review by erisu — APPROVED** [review]

**Comment by oofaish** [comment]
thanks @erisu - changed the variable name too

**Comment by oofaish** [comment]
Can someone with write access merge this PR? 🤗

**Comment by erisu** [comment]
> Can someone with write access merge this PR? 🤗

I will merge it in by tomorrow (JST).
I am waiting for about 12hrs to give other PMC members a chance to leave a review.

**Comment by oofaish** [comment]
🙏

**Review by breautek — APPROVED** [review]
I'm not in a position to be able to physically test these changes as I don't have access to a mac, so I'll be relying on community members to make that distinction...

But codewise everything appears okay.

**Comment by matthew2564** [comment]
Will this fix be getting published via NPM? The latest version there is 3.0.2 - [cordova-plugin-screen-orientation](https://github.com/apache/cordova-plugin-screen-orientation).

**Comment by oofaish** [comment]
@erisu do I need to create a new tag? can we do this manually or should I be putting up another PR?

**Comment by breautek** [comment]
> Will this fix be getting published via NPM? The latest version there is 3.0.2 - [cordova-plugin-screen-orientation](https://www.npmjs.com/package/cordova-plugin-screen-orientation).

It will be, but we cannot provide an ETA.

> do I need to create a new tag? can we do this manually or should I be putting up another PR?

Nope, the tags are all managed by the release processed.

**Comment by shruti-talekar** [comment]
@oofaish - I tested this solution on iPhone 13 Pro Max v16.1. But the fix is not working. I am getting black(BLANK) screen when device is rotated to Landscape mode. What could be the issue? Can you please help

**Comment by oofaish** [comment]
@shruti-talekar please open a new issue and describe the problem, how you compiled the code, any error messages, etc - my tests have been working fine I am afraid.

**Comment by livesemantics** [comment]
Great job guys! How do I install this fix if it's not published to npm (I'm using Capacitor 4.6)? 
I tried to npm install directly from github but keep getting error messages (-4058). maybe I'm doing it wrong...

**Comment by breautek** [comment]
> Great job guys! How do I install this fix if it's not published to npm (I'm using Capacitor 4.6)? I tried to npm install directly from github but keep getting error messages (-4058). maybe I'm doing it wrong...

I'm not familiar with Capacitor, but `npm install` is usually not enough, it simply installs the package into node_modules, but Cordova doesn't know about it and doesn't actually installs it into your platforms.

With the Cordova CLI, you'd use the git url with `cordova plugin add` like so:

```
cordova plugin remove cordova-plugin-screen-orientation
cordova plugin add https://github.com/apache/cordova-plugin-screen-orientation.git
```

The first command will remove the existing plugin install so that you can reinstall the newer version. Note that installing from git are unreleased versions not suitable for production.

I'm not sure how Capacitor exposes these CLI actions, so you may have to resort to their documentation.

**Comment by livesemantics** [comment]
> > Great job guys! How do I install this fix if it's not published to npm (I'm using Capacitor 4.6)? I tried to npm install directly from github but keep getting error messages (-4058). maybe I'm doing it wrong...
> 
> I'm not familiar with Capacitor, but `npm install` is usually not enough, it simply installs the package into node_modules, but Cordova doesn't know about it and doesn't actually installs it into your platforms.
> 
> With the Cordova CLI, you'd use the git url with `cordova plugin add` like so:
> 
> ```
> cordova plugin remove cordova-plugin-screen-orientation
> cordova plugin add https://github.com/apache/cordova-plugin-screen-orientation.git
> ```
> 
> The first command will remove the existing plugin install so that you can reinstall the newer version. Note that installing from git are unreleased versions not suitable for production.
> 
> I'm not sure how Capacitor exposes these CLI actions, so you may have to resort to their documentation.

Capacitor is not that much different. You simply run `npm install` and then `npx cap synch`. 
The problem is that I get an error when I run npm install directly from github.

I've tried:
`npm install https://github.com/apache/cordova-plugin-screen-orientation.git `
(as well as other variations) and I keep getting an errors.

**Comment by breautek** [comment]
I'd advise seeing if you can reproduce the error in a pure cordova sample project, and if so, create a new issue noting you're testing the current development version.

The sample project can also be included in the issue.

**Comment by livesemantics** [comment]
> project

Thanks @breautek but I fail to see how this issue has anything to do with either cordova or capacitor. 
It's just a simple npm install error.

If I run `npm install cordova-plugin-screen-orientation` then I get the old plugin (3.02) installed with no errors.
If I run `npm install https://github.com/apache/cordova-plugin-screen-orientation.git` then I get the following error:

npm ERR! code ENOENT
npm ERR! syscall spawn git
npm ERR! path git
npm ERR! errno -4058
npm ERR! enoent An unknown git error occurred
npm ERR! enoent **This is related to npm not being able to find a file**.
npm ERR! enoent

**Comment by breautek** [comment]
```
npm ERR! code ENOENT
npm ERR! syscall spawn git
```

is the error where NPM cannot find git, so it seems you don't have `git` installed, or at the very least the executable isn't found in your system `PATH` variable.

**Comment by livesemantics** [comment]
> variable

Well, I'm using a cloud service to build so I don't have much control over system PATH etc... I just know that it fails to install.
I'm also not so sure that it's "off topic" because plugins are typically published via npm and since you guys went through all the trouble of fixing but stopped short of publishing a new version ("we cannot provide an ETA") then for me at least, it looks like this fix doesn't really help...

**Pull request merged**
